### PR TITLE
ST6RI-540 Space modeling additions

### PR DESCRIPTION
**Pull request by conradbock**
These are the remaining space modeling capabilities (except bounding box orientation, taking as a separate issue for later).  Slides at [ST4MD-392](https://openmbee.atlassian.net/browse/ST4MD-392?focusedCommentId=13514) (May 16, 2022).

**Occurrences.kerml**

Added inner and outer occurrence references nested under Occurrence::spaceBoundary.  These are space slices that also have no spaceBoundary, where the outer one surrounds the inner ones (see SurroundedBy below).  Also added inverse constraint to spaceBoundaryOf.

Added associations from Occurrence to itself (indirectly) specializing OutsideOf:

- JustOutsideOf linking occurrences that have space slices with no space between them (MatesWith).

- MatesWith specializing JustOutsideOf linking occurrences that have no space between them.

- InnerSpaceOf linking an occurrence to another that completely occupies the space surrounded by an inner space boundary of the first occurrence.  Added Occurrence::innerSpaceOccurrences as cross-navigation feature for this.

- SurroundedBy linking an occurrence that is included in space by an innerSpaceOccurrence of the other.  Types additional connectors in Occurrence that encure spaceBoundary surrounds spaceInterior, and spaceInterior surrounds inner space boundaries.

<br/>

**Items.sysml**

Generalized Item::envelopingShapes to cover curves and surfaces enveloping 2 dimensional shapes.  Nested original constraint.

Added Item::boundingShapes specializing envelopingShapes as StructuredSpaceObjects in which every face/edge intersects the Item.

Added Item::voids as SysML name for innerSpaceOccurrences.  Defined isSolid as having no voids.

<br/>

**CarWithEnvelopingShape.sysml**

Replaced envelopingShapes with boundingShapes.

<br/>

**ShapeItems.sysml**

Minor updates.

### CODE REVIEW

File: sysml.library/Systems Library/Items.sysml [hunk 1]
```diff
@@ -29,29 +35,53 @@ package Items {
...
+		 * Each enveloping shape is the shape of an Item that includes this Item in space and time.
 		 */
-		item envelopingShapes : Item[0..*] {
-			attribute :>> innerSpaceDimension = shape.innerSpaceDimension;
+		item envelopingShapes : Item[0..*] {  /* envelopingBoundaries?	Occurrence::enclosed => included, envelopingBoundaries -> enclose ?*/
+
+			  /* Enables 2 dim items to be enveloped by 2 or 3 dim shapes. */
+			attribute :>> innerSpaceDimension = (Item::innerSpaceDimension == 3  | Item::outerSpaceDimension == 3
+							     ? 2 : Item::outerSpaceDimension-1);   /* Curves bounding flat surfaces SHOULD BE COPLANAR ? */
+			assert constraint { Item::innerSpaceDimension < 3 implies notEmpty(outerSpaceDimension) }
```
Comment (seidewitz) [inline sysml.library/Systems Library/Items.sysml:45]: `/* envelopingBoundaries?	Occurrence::enclosed => included, envelopingBoundaries -> enclose ?*/`
`/* Curves bounding flat surfaces SHOULD BE COPLANAR ? */`

These questions need to be resolved or the comments removed.

File: sysml.library/Systems Library/Items.sysml [hunk 2]
```diff
@@ -67,9 +97,16 @@ package Items {
 		 * Constraints that have been checked by this item.
 		 */
 		abstract ref constraint checkedConstraints: ConstraintCheck[0..*] :> constraintChecks, enactedPerformances;
-
 	}
 	
+	/**
+	 * Touching occurrences are just outside each other and happen at the same time.
+	 */
```
Comment (seidewitz) [inline sysml.library/Systems Library/Items.sysml:104]: "Touching occurrences" should be "Touching items".

**Review by seidewitz — CHANGES_REQUESTED** [review]

**Comment by seidewitz** [comment]
@hpdekoning
Will you be able to review this?

File: sysml.library/Systems Library/Items.sysml (hunk 2 above)
Comment (conradbock) [reply sysml.library/Systems Library/Items.sysml:104]: K, will fix along with previous comment (and commit after a bit to let more feedback come in).

File: sysml.library/Systems Library/Items.sysml (hunk 1 above)
Comment (conradbock) [reply sysml.library/Systems Library/Items.sysml:45]: Was thinking we might want to have fewer "en" words (enclose, envelop), maybe by using them only in SysML and changing "enclosed" to "included" in Kernel (eg, timeIncludedOccurrences).  Then envelopingShapes would be enclosingShapes.

**Comment by conradbock** [comment]
Should I update the various `::self` to `that` (if it's being added in this ST6 sprint)?  Eg, in the Items::`envelopingShapes` constraint Item::`self` would be `that.that`.

File: sysml.library/Systems Library/Items.sysml (hunk 1 above)
Comment (seidewitz) [reply sysml.library/Systems Library/Items.sysml:45]: Not worth worrying more about naming now. Time to move on.

**Comment by seidewitz** [comment]
> Should I update the various `::self` to `that` (if it's being added in this ST6 sprint)? Eg, in the Items::`envelopingShapes` constraint Item::`self` would be `that.that`.

Not yet, since `that` isn't there yet. We can make revisions when once it is introduced. But let's get this PR done first.

File: sysml.library/Systems Library/Items.sysml (hunk 2 above)
Comment (seidewitz) [reply sysml.library/Systems Library/Items.sysml:104]: Please make any updates to branch ST6RI-540 now. I don't want to hold up the PR any longer.

**Comment by conradbock** [comment]
Looks like after all that work on JustOutsideOf I forgot to use it to replace WithinBoth in ShapeItems.  Can I do this now?  Very easy, just a text replace.

**Comment by seidewitz** [comment]
> Looks like after all that work on JustOutsideOf I forgot to use it to replace WithinBoth in ShapeItems. Can I do this now? Very easy, just a text replace.

Sure, go ahead. I was going to merge this PR this morning, but didn't get a chance.

**Comment by seidewitz** [comment]
> can update in TWC ST4MD-392 branch if you haven't merged it yet.

ST4MD-392 is not merged yet.

**Pull request merged**